src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -34456,6 +34440,455 @@ public class Derived : Base
+                );
+        }
+
+        [Fact]
+        public void NotNullConstrainedParam_DoesNotUpdateArgumentState()
+        {
+            var source = @"
+public class Program
+{
+    void M0<T>(T t) where T : notnull { }
+
+    void M1(string? s)
+    {
+        M0(s); // 1
+        _ = s.ToString(); // 2
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'Program.M0<T>(T)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
+                //         M0(s); // 1
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "M0").WithArguments("Program.M0<T>(T)", "T", "string?").WithLocation(8, 9),
+                // (9,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = s.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
+        public void NotNullConstrainedParam_SuppressedArgument_DoesNotUpdateArgumentState()
+        {
+            var source = @"
+public class Program
+{
+    void M0<T>(T t) where T : notnull { }
+
+    void M1(string? s)
+    {
+        M0(s!);
+        _ = s.ToString(); // 1
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (9,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = s.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
+        public void Params_NotNullConstrainedElement_AnnotatedArray_DoesNotUpdateArgumentState()
+        {
+            var source = @"
+public class Program
+{
+    void M0<T>(params T[]? s0) where T : notnull { }
+
+    void M1(string? s1)
+    {
+        M0(s1); // 1
+        _ = s1.ToString(); // 2
+    }
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'Program.M0<T>(params T[]?)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
+                //         M0(s1); // 1
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "M0").WithArguments("Program.M0<T>(params T[]?)", "T", "string?").WithLocation(8, 9),
+                // (9,13): warning CS8602: Dereference of a possibly null reference.
+                //         _ = s1.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s1").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
```

Comment:
Consider adding a test here there is an `override` that removes the nullability annotation and calls through both the base and derived type and shows the outcome on the null state of the argument. Basically making sure we are considering the base and derived method annotation as appropriate. 

I actually can't remember off hand if we look at the declaration method or the override when considering annotations hence I can't remember what the correct behavior is here. Hence a test :smile: